Mark HBase RPMs as noarch

charlesconnell opened the pull request:
I've poked around in the HBase repo and could not find any instance where native code is compiled as part of the HBase build process. I've poked around in `/usr/lib/hbase` on one of our RegionServers, and could not find any native code installed directly into the filesystem.

I did find a number of dependency JARs that provide native libraries inside them. All of these except one provide numerous builds so they work on a variety of architectures. The one exception is the brotli4j library, which only offers native support for x86-64, at least as our build process packages it. We could potentially change our build process to package more brotli JARs to close this gap. I don't believe Hubspot uses the related feature anyways.

I've tested running HBase out of this RPM on my personal cluster and it seems normal. I don't yet have the ability to set up an arm64 RegionServer to see how it behaves there.

charlesconnell committed:
Build HBase as noarch

charlesconnell committed:
use spec file

bbeaudreault commented:
I have not read all of these, but there are a number of HBase JIRAs related to aarch64 and arm64. Here's the aarch64 link: https://issues.apache.org/jira/browse/HBASE-19547?jql=project%20%3D%20HBASE%20AND%20text%20~%20aarch64. Just change the last keyword to arm64 to see those.

We should go over those closely for anything that might influence our RPM or install process. I see something related to netty epoll for example, which could be a big issue if it didn't perform well (but again did not read any of it closely).

charlesconnell commented:
Alright let's go through these:
- HBASE-25170	supporting hbase on aarch64 platform
  - Claims that leveldbjni dependency contains native code without aarch64 support. I ran `unzip -l leveldbjni-all-1.8.jar` to see for myself. The relevant lines are
    ```
        0  10-17-2013 09:18   META-INF/native/
        0  10-17-2013 09:18   META-INF/native/linux32/
   699040  10-16-2013 10:01   META-INF/native/linux32/libleveldbjni.so
        0  10-17-2013 09:18   META-INF/native/linux64/
   752803  10-16-2013 09:53   META-INF/native/linux64/libleveldbjni.so
        0  10-17-2013 09:18   META-INF/native/osx/
   480504  10-16-2013 12:31   META-INF/native/osx/libleveldbjni.jnilib
        0  10-17-2013 09:18   META-INF/native/windows32/
   225792  10-17-2013 08:53   META-INF/native/windows32/leveldbjni.dll
        0  10-17-2013 09:18   META-INF/native/windows64/
   267264  10-17-2013 09:14   META-INF/native/windows64/leveldbjni.dll
    ```
    I guess I missed this on my earlier search because it doesn't even use "x86" in the file names. If I take "linux64" to mean x86-64 then this could be an issue. More comments below.
- HBASE-19547	HBase fails building on AArch64 due to asciidoctor-maven-plugin.
  - Fixed, and we're not necessarily concerned with build problems
- HBASE-24266	Document of Hbase on Aarch64
  - Documentation only, and not interesting at that
- HBASE-26602	[aarch64] TestAsyncNonMetaRegionLocator is flaky
  - cannot reproduce, and affects test only
- HBASE-24178	some tests of org.apache.hadoop.hbase.TestZooKeeper.* failed on aarch64
  - just affects unit tests
- HBASE-26600	[branch-2, create-release] Private per build Maven caches won't work on aarch64 MacOS
  - build issue, not important
- HBASE-27113	add Netty epoll support for aarch64
  - could be significant
- HBASE-25565	test fails on AArch64, Fedora 33, but build worked
  - just affects unit tests
- HBASE-19146	Hbase3.0  protobuf-maven-plugin do not support Arm64(only for x86)
  - not relevant to 2.x
- HBASE-26563	TestCacheOnWrite and TestCompactingToCellFlatMapMemStore use too many file descriptors
  - just affects unit tests, and unclear if this is an aarch64 specific-issue at all anyways
- HBASE-26761	TestMobStoreScanner (testGetMassive) can OOME
  - fixed, and wasn't aarch64-specific
- HBASE-26492	[branch-2, hbase-thirdparty, Java 8] TestUnloadAccessController and other unit tests fail to start due to ByteBuffer link error
  - fixed, and wasn't aarch64-specific
- HBASE-27103	All MR UTs are broken because of ClassNotFound
  - fixed
- HBASE-27065	Build against Hadoop 3.3.3
  - More of leveldbjndi issue, see my comments below
- HBASE-27054	TestStochasticLoadBalancerRegionReplicaLargeCluster.testRegionReplicasOnLargeCluster is flaky  
  - Not aarch64-specific
- HBASE-26844	Fix flaky TestBasicWALEntryStreamFSHLog.testSizeOfLogQueue
  - only affects tests
- HBASE-26983	Upgrade JRuby to 9.3.4.0
  - Claims that `hbase shell` doesn't work on aarch64 unless you get the new JRuby. This ticket is resolved but not in 2.5.x, so we might really have a problem here.
- HBASE-27101	support commons-crypto version 1.1.0
  - fixed in 2.5.0
- HBASE-22792	using ASYNC_WAL in hbase sequentialWrite, but RS crashes
  - Not likely an architecture-specific issue
- HBASE-20459	Majority of scan CPU time in HBase-1 spent in size estimation
  - Not likely an architecture-specific issue
- HBASE-27106	HBase site generation should use its own JRuby
  - Does not affect 2.x
- HBASE-19387	HBase-spark snappy.SnappyError on Arm64
  - Only affects `hbase-spark` module
- HBASE-27075	TestUpdateRSGroupConfiguration.testCustomOnlineConfigChangeInRSGroup is flaky
  - Not an architecture-specific issue
- HBASE-27072	TestCoprocessorEndpointTracing.traceAsyncTableEndpoint is flaky
  - Not likely an architecture-specific issue
- HBASE-27071	TestMemStoreLAB is flaky
  - Not likely an architecture-specific issue
- HBASE-27074	TestProcedureSchedulerConcurrency.testConcurrentWaitWake is flaky
  - Not likely an architecture-specific issue
- HBASE-27073	TestReplicationValueCompressedWAL.testMultiplePuts is flaky
  - Not likely an architecture-specific issue
- HBASE-26493	[branch-2] TestLruBlockCache is flaky
  - Not likely an architecture-specific issue
- HBASE-26717	[hbase-thirdparty] hbase-shaded-jersey packages unshaded classes
  - Not likely an architecture-specific issue
- HBASE-26541	hbase-protocol-shaded not buildable on M1 MacOSX
  - Build-only problem, but could affect us if we do an aarch64 build
- HBASE-26825	TestBasicWALEntryStreamFSHLog testSizeOfLogQueue is flaky
  - Not likely an architecture-specific issue
- HBASE-19386	HBase UnsafeAvailChecker returns false on Arm64
  - fixed
- HBASE-19398	Unit Tests failed in hbase-3.0.0 on Arm64 and x86
  - Unclear how this is an aarch64 problem
- HBASE-26470	Use openlabtesting protoc on linux arm64 in HBASE 2.x
 - Fixed, and besides we have our own protoc aarch64 build now
- HBASE-19146	Hbase3.0  protobuf-maven-plugin do not support Arm64(only for x86)
  - We already know about protoc stuff
- HBASE-27338	brotli compression lib tests fail on arm64
  - We know about this potential issue already
- HBASE-23612	Update pom.xml to use another 2.5.0 protoc as external protobuf
  - We already know about protoc stuff
- HBASE-24259	[thirdparty] Update Protobuf to 3.11.4
  - I believe this affects 3.x only
- HBASE-27167	s390x: Skip tests on unsupported compression libs
  - Not actually about aarch64

Summary of issues:
- `leveldbjni` aarch64 support
  - There is a different JAR that provides aarch64 support for leveldb. If we built an aarch64 RPM I believe we could have it in there.
- netty+epoll support
  - Investigation required
- hbase shell / JRuby support
  - Investigation required

bbeaudreault commented:
Thanks for summarizing. Do you want to move the investigation into your main issue then? I don't mind merging this since it seems relatively harmless for now, but we should probably investigate those before long.

charlesconnell commented:
## `leveldbjni`

According to `mvn dependency:tree -Phadoop-3.0 -Dhadoop.profile=3.0 -Dhadoop-three.version=3.3.1`, this is in HBase's dependency tree because of
```
[INFO] +- org.apache.hadoop:hadoop-hdfs:test-jar:tests:3.3.1:test
[INFO] |  +- com.sun.jersey:jersey-server:jar:1.19:compile
[INFO] |  +- io.netty:netty:jar:3.10.6.Final:test
[INFO] |  +- io.netty:netty-all:jar:4.1.61.Final:test
[INFO] |  \- org.fusesource.leveldbjni:leveldbjni-all:jar:1.8:test
```
so it seems wrong that it would get into the RPM at all, anyways. We must be including test-scope stuff in the RPM. The code agrees that it's used in tests exclusively. The only code references to `leveldb` are in `AsyncFSTestBase` and `HBaseTestingUtility`. I think it's worth merging this PR and seeing how it goes when we launch on an aarch64 machine, with respect to this issue at least.

## Netty epoll

According to https://netty.io/news/2020/05/13/4-1-50-Final.html, Netty version 4.1.50 "supports AARCH64 for the native epoll transport." HBase 2.5.0 uses a repackaged version of netty, hbase-shaded-netty 4.1.3. This is a repackaging of [netty 4.1.84](https://github.com/apache/hbase-thirdparty/blob/rel/4.1.3/pom.xml#L136), so it's new enough. This agrees with what I saw on a RegionServer before opening this PR:
```
 QA - test01  [cconnell@test2-alive-snail hbase]$ unzip -l hbase-shaded-netty-4.1.3.jar | fgrep '.so'
   103728  10-11-2022 23:24   META-INF/native/liborg_apache_hbase_thirdparty_netty_transport_native_epoll_aarch_64.so
    94722  10-11-2022 23:27   META-INF/native/liborg_apache_hbase_thirdparty_netty_transport_native_epoll_x86_64.so
```

## JRuby too old

This stands out a potentially significant issue because we really do want `hbase shell` to work. However, applying the change in this PR won't make it any better or worse.

charlesconnell commented:
Running this now, I did some basic tinkering with `hbase shell`, and it seems fine
